### PR TITLE
Feature: transform the display

### LINKED ISSUE #41: Feature: Transformation
I am using this on river window manager, Do I need to use kanshi on top of way-display to rotate the diplay or is there any option that I cannot seem to find?

### PR DESCRIPTION
**Pull request by alex-courtis**
Closes #41 

@anandubey let's continue discussions in here.

### CODE REVIEW

File: src/cfg.cpp [hunk 1]
```diff
@@ -516,6 +572,24 @@ void cfg_parse_node(struct Cfg *cfg, YAML::Node &node) {
 		}
 	}
 
+	if (node["TRANSFORM"]) {
+		for (const auto &transform : node["TRANSFORM"]) {
+			struct UserTransform *user_transform = cfg_user_transform_default();
+
+			if (!parse_node_val_string(transform, "NAME_DESC", &user_transform->name_desc, "TRANSFORM", "")) {
+				cfg_user_mode_free(user_transform);
```
Comment (alex-courtis) [inline src/cfg.cpp:580]: Should be `cfg_user_transform_free`

File: src/cfg.cpp [hunk 2]
```diff
@@ -516,6 +572,24 @@ void cfg_parse_node(struct Cfg *cfg, YAML::Node &node) {
...
+				continue;
+			}
+			if (!parse_node_val_int(transform, "DEGREE", &user_transform->transform, "TRANSFORM", user_transform->name_desc)) {
```
Comment (alex-courtis) [inline src/cfg.cpp:583]: We will need to be more strict about the value here: only a few values are applicable, and we need to account for flipped.

We can use a similar scheme to [sway](https://man.archlinux.org/man/sway-output.5.en):

> Can be one of "90", "180", "270" for rotation; or "FLIPPED", "FLIPPED-90", "FLIPPED-180", "FLIPPED-270" to apply a rotation and flip, or "NORMAL" to apply no transform.

convert.c methods similar to `auto_scale_val` and `auto_scale_name` will need to be created to perform that validation and conversion.

File: inc/cfg.h [hunk 3]
```diff
@@ -52,6 +52,11 @@ struct UserMode {
 	bool warned_no_mode;
 };
 
+struct UserTransform {
+	char *name_desc;
+	int32_t transform;
```
Comment (alex-courtis) [inline inc/cfg.h:57]: This should be a `wl_output_transform`

File: src/cfg.cpp [hunk 4]
```diff
@@ -164,6 +179,25 @@ bool equal_user_mode(const void *value, const void *data) {
...
 }
 
+bool equal_user_transform(const void *value, const void *data) {
+	if (!value || !data) {
+		return false;
+	}
+
+	struct UserTransform *lhs = (struct UserTransform*)value;
+	struct UserTransform *rhs = (struct UserTransform*)data;
+
```
Comment (alex-courtis) [inline src/cfg.cpp:189]: We only need to test the values for equality. No need to look at names.

There is no such thing as a "null" transform, as the default `WL_OUTPUT_TRANSFORM_NORMAL` is 0.

File: src/head.c [hunk 5]
```diff
@@ -182,12 +182,21 @@ void head_scaled_dimensions(struct Head *head) {
 		return;
 	}
 
-	if (head->transform % 2 == 0) {
+	if (head->transform == WL_OUTPUT_TRANSFORM_NORMAL ||
```
Comment (alex-courtis) [inline src/head.c:185]: `head->transform % 2 == 0` is actually enough, as multiples of 90 are odd values.

File: src/layout.c [hunk 6]
```diff
@@ -240,6 +249,7 @@ void apply(void) {
 				head->zwlr_config_head = zwlr_output_configuration_v1_enable_head(zwlr_config, head->zwlr_head);
 				zwlr_output_configuration_head_v1_set_scale(head->zwlr_config_head, head->desired.scale);
 				zwlr_output_configuration_head_v1_set_position(head->zwlr_config_head, head->desired.x, head->desired.y);
+				zwlr_output_configuration_head_v1_set_transform(head->zwlr_config_head, head->desired.transform);
```
Comment (alex-courtis) [inline src/layout.c:252]: This definitely works!

**Review by alex-courtis — COMMENTED** [review]
This is looking good and the 180 flip is working.

TODO

- [x] fixes mentioned in review
- [x] Actually apply the Cfg value in `desire_transform`
- [x] Print the transformation information in `info.c`
- [ ] Add transformation to CLI; merge etc. looks complete so I expect it should Just Work

File: src/cfg.cpp (hunk 1 above)
Comment (anandubey) [reply src/cfg.cpp]: Done.

File: src/cfg.cpp (hunk 2 above)
Comment (anandubey) [reply src/cfg.cpp]: Need time to figure out this one. Not getting enough time from work :(

File: inc/cfg.h (hunk 3 above)
Comment (anandubey) [reply inc/cfg.h]: Done.

File: src/cfg.cpp (hunk 4 above)
Comment (alex-courtis) [reply src/cfg.cpp:189]: I realise I have made a huge mistake; we do need to test the names and values in this test as it is for the entire config.

Like `equal_user_mode`

```c
bool equal_user_transform(const void *value, const void *data) {
	if (!value || !data) {
		return false;
	}

	struct UserTransform *lhs = (struct UserTransform*)value;
	struct UserTransform *rhs = (struct UserTransform*)data;

	if (!lhs->name_desc || !rhs->name_desc) {
		return false;
	}

	if (strcasecmp(lhs->name_desc, rhs->name_desc) != 0) {
		return false;
	}

	if (lhs->transform != rhs->transform) {
		return false;
	}

	return true;
}
```

File: src/cfg.cpp [hunk 7]
```diff
@@ -612,6 +682,22 @@ void cfg_emit(YAML::Emitter &e, struct Cfg *cfg) {
 		e << YAML::EndSeq;
 	}
 
+	if (cfg->user_transform) {
+		e << YAML::Key << "TRANSFORM";
+		e << YAML::BeginSeq;
+		for (struct SList *i = cfg->user_modes; i; i = i->nex) {
```
Comment (alex-courtis) [inline src/cfg.cpp:688]: Should be cfg->user_transforms

**Review by alex-courtis — COMMENTED** [review]
Made a few fixes. I don't have access you push to your repo, so here are some patches you can `git apply ...` in order.
[0001-formatting.patch.txt](https://github.com/alex-courtis/way-displays/files/9081000/0001-formatting.patch.txt)
[0002-remove-head.tranform-always-use-desired-and-current.patch.txt](https://github.com/alex-courtis/way-displays/files/9081002/0002-remove-head.tranform-always-use-desired-and-current.patch.txt)
[0003-assorted-fixes-for-transform-read-write.patch.txt](https://github.com/alex-courtis/way-displays/files/9081003/0003-assorted-fixes-for-transform-read-write.patch.txt)


This is working quite nicely now.

Tested OK:
* value from cfg.yaml
* set and delete from CLI
* write from CLI

TODO:
- [ ] user friendly strings -> enums https://github.com/alex-courtis/way-displays/pull/43#discussion_r889784274
- [ ] doc: readme and man page
- [ ] consistency in naming: always use "transform", not "rotation"

**Comment by alex-courtis** [comment]
There is a large refactor incoming, which greatly affects cfg.

I am happy to merge it to your repo.

I'll need to have push access... applying a large merge as a patch is not viable.

**Comment by alex-courtis** [comment]
> I have added you to the collaborators list. I hope now you will have the push access.

Thanks @anandubey 

I'll merge master to your fork after https://github.com/alex-courtis/way-displays/tree/49-yaml-schemas is merged.

**Comment by alex-courtis** [comment]
> I'll merge master to your fork after https://github.com/alex-courtis/way-displays/tree/49-yaml-schemas is merged.

master at 1.6.0 has been merged, along with the patches above.

The large change resulted from refactors when adding the [IPC API](https://github.com/alex-courtis/way-displays/blob/master/doc/IPC.md). A lot of src/cfg was moved to src/marshalling

I've pre-emptively added !!transform to the [YAML schema](https://github.com/alex-courtis/way-displays/blob/master/doc/YAML_SCHEMAS.md#transform)
You'll need test it with the [example client](https://github.com/alex-courtis/way-displays/blob/master/doc/IPC.md) when you have finished adding all the TRANSFORM enums.

**Comment by alex-courtis** [comment]
Nudge @anandubey 

Any updates on this one?

This is really close to complete and I look forward to releasing it!

**Comment by anandubey** [comment]
@alex-courtis Sorry for zoning out on this. Started my new job on 1st Aug. It's a rust dev role, so all my weekends are going in that. I will try my best to finish it this week.

**Comment by alex-courtis** [comment]
> @alex-courtis Sorry for zoning out on this. Started my new job on 1st Aug. It's a rust dev role, so all my weekends are going in that. I will try my best to finish it this week.

No worries. There is no hurry at all.

Good luck in the new role!

**Comment by edrex** [comment]
Thanks for your work on this @anandubey, and totally understand about prioritizing paid job. 

It would be great to get output transforms in tho. Do you anticipate having time to get this merge ready going forward, or alternatively, would you be open to handing it off?

If C was fresher in my mind I would volunteer (and my ADHD brain likes to say yes) but I shouldn't.

**Comment by alex-courtis** [comment]
> would you be open to handing it off?

That would be fantastic!

I just cannot allocate the time to finish this one off...

**Comment by edrex** [comment]
> make cppcheck
cool!

> way-displays is also set up for ccls live checking in the editor.
alright! I use helix which [uses](https://docs.helix-editor.com/lang-support.html) clangd by default but I can change it in `languages.toml`.

**Comment by alex-courtis** [comment]
> alright! I use helix which [uses](https://docs.helix-editor.com/lang-support.html) clangd by default but I can change it in languages.toml.

All good; clang is preferred. gcc is only used by default for legacy packaging reasons.

**Comment by alex-courtis** [comment]
Added development information to https://github.com/alex-courtis/way-displays/blob/master/CONTRIBUTING.md

**Comment by edrex** [comment]
Want to try to get this in before #83? I would be willing to do a testing pass and make any needed changes if you have time for code review. I think I basically remember how to C :)

I'll be on a road trip/camping Apr 14-23 though, so we'll see how much keyboard time I get.

**Comment by alex-courtis** [comment]
> Want to try to get this in before https://github.com/alex-courtis/way-displays/issues/83? I would be willing to do a testing pass and make any needed changes if you have time for code review. I think I basically remember how to C :)
> 
> I'll be on a road trip/camping Apr 14-23 though, so we'll see how much keyboard time I get.

Oh yes please!

I've added a lot of unit tests on master, so that should make this easier RE merge etc.

**Comment by alex-courtis** [comment]
#133

**Pull request closed without merging**